api/oneclick.py

```diff
@@ -419,49 +417,78 @@ def internal_format(self, delivery_mechanism):
+            if not remote_identifier:
+                remote_identifier = self.create_patron(patron)
+            credential.credential = remote_identifier
+            credential.expires = None
+
+        _db = Session.object_session(patron)
+        credential = Credential.lookup(
+            _db, DataSource.RB_DIGITAL,
+            Credential.IDENTIFIER_FROM_REMOTE_SERVICE,
+            patron, refresher_method=refresher,
+            allow_persistent_token=True
+        )
+        if not credential.credential:
+            refresher(credential)
+        return credential.credential
+
+    def create_patron(self, patron):
+        """Ask RBdigital to create a new patron record.
 
-        # TODO: will not work in all libraries, find a better solution
-        patron_cardno = patron.authorization_identifier
-        if not patron_cardno:
-            raise InvalidInputException("Patron %r has no card number.", patron)
+        :param patron: the Patron that needs a new RBdigital account.
+
+        :return The internal RBdigital identifier for this patron.
+        """
 
         url = "%s/libraries/%s/patrons/" % (self.base_url, str(self.library_id))
         action="create_patron"
-        
+
+        patron_identifier = patron.identifier_to_remote_service(
+            DataSource.RB_DIGITAL
+        )
+
         post_args = dict()
         post_args['libraryId'] = self.library_id
-        post_args['libraryCardNumber'] = str(patron.authorization_identifier)
-        # generate random values for the account fields the patron has not supplied us with
-        patron_uuid = str(uuid.uuid1())
-        post_args['userName'] = 'username_' + patron_uuid
-        post_args['email'] = 'patron_' + patron_uuid + '@librarysimplified.org'
+        post_args['libraryCardNumber'] = patron_identifier
+
+        # Generate meaningless values for account fields that are not
+        # relevant to our usage of the API.
+        post_args['userName'] = 'username_' + patron_identifier
+        post_args['email'] = 'patron_' + patron_identifier + '@librarysimplified.org'
         post_args['firstName'] = 'Patron'
         post_args['lastName'] = 'Reader'
-        # will not be used in our system, so just needs to be set to a securely randomized value
-        post_args['password'] = os.urandom(8).encode('hex')
 
+        # The patron will not be logging in to this RBdigital account,
+        # so set their password to a secure value and forget it.
+        post_args['password'] = os.urandom(8).encode('hex')
 
         resp_dict = {}
         message = None
-        try:
-            response = self.request(url=url, data=json.dumps(post_args), method="post")
-            if response.text:
-                resp_dict = response.json()
-                message = resp_dict.get('message', None)
-        except Exception, e:
-            self.log.error("Patron create failed: %r", e, exc_info=e)
-            raise RemoteInitiatedServerError(e.message, action)
```

Comment:
I'm not sure about this but I think it's more useful to raise the actual exception here rather than catching it and reraising a different one.